pandas/core/arrays/datetimelike.py

```diff
@@ -347,21 +326,59 @@ def _validate_frequency(cls, index, freq, **kwargs):
     # Arithmetic Methods
 
     def _add_datelike(self, other):
+        # Overriden by TimedeltaArray
         raise TypeError("cannot add {cls} and {typ}"
                         .format(cls=type(self).__name__,
                                 typ=type(other).__name__))
 
+    _add_datelike_dti = _add_datelike
+
     def _sub_datelike(self, other):
-        raise com.AbstractMethodError(self)
+        # Overridden by DatetimeArray
+        assert other is not NaT
+        raise TypeError("cannot subtract a datelike from a {cls}"
+                        .format(cls=type(self).__name__))
+
+    _sub_datelike_dti = _sub_datelike
 
     def _sub_period(self, other):
-        return NotImplemented
+        # Overriden by PeriodArray
+        raise TypeError("cannot subtract Period from a {cls}"
+                        .format(cls=type(self).__name__))
 
     def _add_offset(self, offset):
         raise com.AbstractMethodError(self)
 
     def _add_delta(self, other):
-        return NotImplemented
+        """
+        Add a timedelta-like, Tick or TimedeltaIndex-like object
+        to self, yielding another array of the same type
+
+        Parameters
+        ----------
+        delta : {timedelta, np.timedelta64, Tick,
+                 TimedeltaIndex, ndarray[timedelta64]}
+
+        Returns
+        -------
+        result : same type as self
+
+        Notes
+        -----
+        The result's name is set outside of _add_delta by the calling
+        method (__add__ or __sub__), if necessary (i.e. for Indexes).
+        """
+        # Note: The docstring here says the return type is the same type
```

Comment:
this is a very confusing note here. As a reader I wouldn't know what to make of it. I would just list the possible return values here, or say same as sub-class type.

Reply:
Will make this explicit.